src/services/codefixes/importFixes.ts

```diff
@@ -296,63 +295,99 @@ namespace ts.codefix {
+    function getModuleSpecifiersForNewImport(
+        sourceFile: SourceFile,
+        moduleSymbol: Symbol,
+        options: CompilerOptions,
+        getCanonicalFileName: (file: string) => string,
+        host: LanguageServiceHost,
+    ): string[] {
+        const { baseUrl, paths, rootDirs } = options;
         const moduleFileName = moduleSymbol.valueDeclaration.getSourceFile().fileName;
         const sourceDirectory = getDirectoryPath(sourceFile.fileName);
 
-        return tryGetModuleNameFromAmbientModule(moduleSymbol) ||
-            tryGetModuleNameFromTypeRoots(options, host, getCanonicalFileName, moduleFileName) ||
-            tryGetModuleNameAsNodeModule(options, moduleFileName, host, getCanonicalFileName, sourceDirectory) ||
-            tryGetModuleNameFromBaseUrl(options, moduleFileName, getCanonicalFileName) ||
-            options.rootDirs && tryGetModuleNameFromRootDirs(options.rootDirs, moduleFileName, sourceDirectory, getCanonicalFileName) ||
-            removeFileExtension(getRelativePath(moduleFileName, sourceDirectory, getCanonicalFileName));
-    }
+        const global = tryGetModuleNameFromAmbientModule(moduleSymbol)
+            || tryGetModuleNameFromTypeRoots(options, host, getCanonicalFileName, moduleFileName)
+            || tryGetModuleNameAsNodeModule(options, moduleFileName, host, getCanonicalFileName, sourceDirectory)
+            || rootDirs && tryGetModuleNameFromRootDirs(rootDirs, moduleFileName, sourceDirectory, getCanonicalFileName);
+        if (global) {
+            return [global];
+        }
 
-    function tryGetModuleNameFromAmbientModule(moduleSymbol: Symbol): string | undefined {
-        const decl = moduleSymbol.valueDeclaration;
-        if (isModuleDeclaration(decl) && isStringLiteral(decl.name)) {
-            return decl.name.text;
+        const relativePath = removeFileExtension(getRelativePath(moduleFileName, sourceDirectory, getCanonicalFileName));
+        if (!baseUrl) {
+            return [relativePath];
+        }
+
+        const relativeToBaseUrl = getRelativePathIfInDirectory(moduleFileName, baseUrl, getCanonicalFileName);
+        if (!relativeToBaseUrl) {
+            return [relativePath];
         }
+
+        const importRelativeToBaseUrl = removeExtensionAndIndexPostFix(relativeToBaseUrl);
+        if (paths) {
+            const fromPaths = tryGetModuleNameFromPaths(removeFileExtension(relativeToBaseUrl), importRelativeToBaseUrl, paths);
+            if (fromPaths) {
+                return [fromPaths];
+            }
+        }
+
+        // Prefer a relative import over a baseUrl import if it doesn't traverse up to baseUrl.
+        const relativeFirst = getRelativePathNParents(relativePath) < getRelativePathLength(getRelativePath(sourceDirectory, baseUrl, getCanonicalFileName));

```

Comment:
I am not sure i understand this condition..
should not this just be `getRelativePathNParents(getRelativePath(sourceDirectory, baseUrl, getCanonicalFileName))` ?

Reply:
Added a comment, tell me if this could be done in some simpler way.